quinn-proto/src/connection/receiver_timestamps.rs

```diff
@@ -0,0 +1,126 @@
+use std::{collections::VecDeque, fmt, time::Instant};
+
+use tracing::warn;
+
+#[derive(Debug, Clone, PartialEq, Eq, Copy)]
+pub struct PacketTimestamp {
+    pub packet_number: u64,
+    pub timestamp: Instant,
+}
+
+impl Default for PacketTimestamp {
+    fn default() -> Self {
+        Self {
+            packet_number: 0,
+            timestamp: Instant::now(),
+        }
+    }
+}
+
+pub struct ReceiverTimestamps {
+    pub data: VecDeque<PacketTimestamp>,
+    max: usize,
+}
+
+impl fmt::Debug for ReceiverTimestamps {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut l = f.debug_list();
+        let mut last: Option<(u64, Instant)> = None;
+        for curr in self.data.iter() {
+            if let Some(last) = last.take() {
+                let s = format!(
+                    "{}..{} diff_micros: {}",
+                    last.0,
+                    curr.packet_number,
+                    curr.timestamp.duration_since(last.1).as_micros(),
+                );
+                l.entry(&s);
+            }
+            let _ = last.insert((curr.packet_number, curr.timestamp));
+        }
+        l.finish()
+    }
+}
+
+impl ReceiverTimestamps {
+    pub(crate) fn new(max: usize) -> Self {
+        Self {
+            data: VecDeque::with_capacity(max),
+            max,
+        }
+    }
+
+    pub(crate) fn add(&mut self, packet_number: u64, timestamp: Instant) {
+        if self.data.len() == self.max {
+            self.data.pop_front();
+        }
+        if let Some(v) = self.data.back() {
+            if packet_number <= v.packet_number {
+                warn!("out of order packets are not supported");
```

Comment:
This will be very noisy on some real networks. It should be `trace` at most, or omitted entirely.